airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -101,15 +123,27 @@ def read_records_from_slice(self, stream_slice: StreamSlice) -> Iterable[Airbyte
             n_skipped = line_no = 0
 
             try:
-                for record in parser.parse_records(self.config, file, self.stream_reader, self.logger, schema):
-                    line_no += 1
-                    if self.config.schemaless:
-                        record = {"data": record}
-                    elif not self.record_passes_validation_policy(record):
-                        n_skipped += 1
-                        continue
-                    record = self.transform_record(record, file, file_datetime_string)
-                    yield stream_data_to_airbyte_message(self.name, record)
+                if self.use_file_transfer:
+                    self.logger.info(f"{self.name}: {file} file-based syncing")
+                    # todo: complete here the code to not rely on local parser
+                    blob_transfer = BlobTransfer()
+                    for record in blob_transfer.write_streams(self.config, file, self.stream_reader, self.logger):
+                        line_no += 1
+                        if not self.record_passes_validation_policy(record):
+                            n_skipped += 1
+                            continue
+                        record = self.transform_record_for_file_transfer(record, file)
+                        yield stream_data_to_airbyte_message(self.name, record, is_file_transfer_message=True)
```

Comment:
Not sure where to put this question but here seemed like a reasonable spot... Just to better understand our state message emitting frequency here:

I think that previously a slice _probably?_ equaled a single file, and state messages would be sent after each slice. Does that mean we send state messages after each file? Or is there another logic governing how often we emit state messages? Thanks. 🙏

Reply:
Yes, from what I understand, we emit state messages after each file. At least for a setup like below with each stream = 1 file.

```
{"type":"LOG","log":{"level":"INFO","message":"Starting syncing SourceSFTPBulk"}}
{"type":"LOG","log":{"level":"INFO","message":"Marking stream simpsons_locations as STARTED"}}
{"type":"TRACE","trace":{"type":"STREAM_STATUS","emitted_at":1730318278059.17,"stream_status":{"stream_descriptor":{"name":"simpsons_locations"},"status":"STARTED"}}}
{"type":"LOG","log":{"level":"INFO","message":"Setting state of SourceSFTPBulk stream to {}"}}
{"type":"LOG","log":{"level":"INFO","message":"Syncing stream: simpsons_locations "}}
{"type":"LOG","log":{"level":"INFO","message":"simpsons_locations: uri='//sftp-testing-for-file-transfer/sftp-folder/simpsons_locations.csv' last_modified=datetime.datetime(2024, 10, 7, 13, 56, 3) mime_type=None file-based syncing"}}
{"type":"LOG","log":{"level":"INFO","message":"Writing file to '/tmp/airbyte-file-transfer/sftp-testing-for-file-transfer/sftp-folder/simpsons_locations.csv' with size: 0.17 MB (0.00 GB), available disk space: 571,439.89 MB (558.05 GB),available memory: 6,374.41 MB (6.23 GB)."}}
{"type":"LOG","log":{"level":"INFO","message":"Starting to read the file"}}
{"type":"LOG","log":{"level":"INFO","message":"Time taken to read the file: 0.51 seconds."}}
{"type":"LOG","log":{"level":"INFO","message":"Starting to write the file locally"}}
{"type":"LOG","log":{"level":"INFO","message":"Time taken to write the file: 0.00 seconds."}}
{"type":"LOG","log":{"level":"INFO","message":"File sftp-testing-for-file-transfer/sftp-folder/simpsons_locations.csv successfully written to /tmp/airbyte-file-transfer."}}
{"type":"LOG","log":{"level":"INFO","message":"Marking stream simpsons_locations as RUNNING"}}
{"type":"TRACE","trace":{"type":"STREAM_STATUS","emitted_at":1730318389636.485,"stream_status":{"stream_descriptor":{"name":"simpsons_locations"},"status":"RUNNING"}}}
{"type":"RECORD","record":{"stream":"simpsons_locations","file":{"file_url":"/tmp/airbyte-file-transfer/sftp-testing-for-file-transfer/sftp-folder/simpsons_locations.csv","bytes":182776,"file_relative_path":"sftp-testing-for-file-transfer/sftp-folder/simpsons_locations.csv","modified":1728330963000,"source_file_url":"//sftp-testing-for-file-transfer/sftp-folder/simpsons_locations.csv"},"emitted_at":1730318389636,"data":{}}}
{"type":"STATE","state":{"type":"STREAM","stream":{"stream_descriptor":{"name":"simpsons_locations"},"stream_state":{"history":{"//sftp-testing-for-file-transfer/sftp-folder/simpsons_locations.csv":"2024-10-07T13:56:03.000000Z"},"_ab_source_file_last_modified":"2024-10-07T13:56:03.000000Z_//sftp-testing-for-file-transfer/sftp-folder/simpsons_locations.csv"}},"sourceStats":{"recordCount":1.0}}}
{"type":"LOG","log":{"level":"INFO","message":"Read 1 records from simpsons_locations stream"}}
{"type":"LOG","log":{"level":"INFO","message":"Marking stream simpsons_locations as STOPPED"}}
{"type":"TRACE","trace":{"type":"STREAM_STATUS","emitted_at":1730318389712.224,"stream_status":{"stream_descriptor":{"name":"simpsons_locations"},"status":"COMPLETE"}}}
{"type":"LOG","log":{"level":"INFO","message":"Finished syncing simpsons_locations"}}
{"type":"LOG","log":{"level":"INFO","message":"SourceSFTPBulk runtimes:\nSyncing stream simpsons_locations 0:01:51.654029"}}
{"type":"LOG","log":{"level":"INFO","message":"Marking stream simpsons_characters as STARTED"}}
{"type":"TRACE","trace":{"type":"STREAM_STATUS","emitted_at":1730318389712.517,"stream_status":{"stream_descriptor":{"name":"simpsons_characters"},"status":"STARTED"}}}
{"type":"LOG","log":{"level":"INFO","message":"Setting state of SourceSFTPBulk stream to {}"}}
{"type":"LOG","log":{"level":"INFO","message":"Syncing stream: simpsons_characters "}}
{"type":"LOG","log":{"level":"INFO","message":"simpsons_characters: uri='//sftp-testing-for-file-transfer/sftp-folder/simpsons_characters.csv' last_modified=datetime.datetime(2024, 10, 7, 13, 56, 3) mime_type=None file-based syncing"}}
{"type":"LOG","log":{"level":"INFO","message":"Writing file to '/tmp/airbyte-file-transfer/sftp-testing-for-file-transfer/sftp-folder/simpsons_characters.csv' with size: 0.21 MB (0.00 GB), available disk space: 571,417.51 MB (558.02 GB),available memory: 6,809.19 MB (6.65 GB)."}}
{"type":"LOG","log":{"level":"INFO","message":"Starting to read the file"}}
{"type":"LOG","log":{"level":"INFO","message":"Time taken to read the file: 0.47 seconds."}}
{"type":"LOG","log":{"level":"INFO","message":"Starting to write the file locally"}}
{"type":"LOG","log":{"level":"INFO","message":"Time taken to write the file: 0.00 seconds."}}
{"type":"LOG","log":{"level":"INFO","message":"File sftp-testing-for-file-transfer/sftp-folder/simpsons_characters.csv successfully written to /tmp/airbyte-file-transfer."}}
{"type":"LOG","log":{"level":"INFO","message":"Marking stream simpsons_characters as RUNNING"}}
{"type":"TRACE","trace":{"type":"STREAM_STATUS","emitted_at":1730318468966.405,"stream_status":{"stream_descriptor":{"name":"simpsons_characters"},"status":"RUNNING"}}}
{"type":"RECORD","record":{"stream":"simpsons_characters","file":{"file_url":"/tmp/airbyte-file-transfer/sftp-testing-for-file-transfer/sftp-folder/simpsons_characters.csv","bytes":222548,"file_relative_path":"sftp-testing-for-file-transfer/sftp-folder/simpsons_characters.csv","modified":1728330963000,"source_file_url":"//sftp-testing-for-file-transfer/sftp-folder/simpsons_characters.csv"},"emitted_at":1730318468966,"data":{}}}
{"type":"STATE","state":{"type":"STREAM","stream":{"stream_descriptor":{"name":"simpsons_characters"},"stream_state":{"history":{"//sftp-testing-for-file-transfer/sftp-folder/simpsons_characters.csv":"2024-10-07T13:56:03.000000Z"},"_ab_source_file_last_modified":"2024-10-07T13:56:03.000000Z_//sftp-testing-for-file-transfer/sftp-folder/simpsons_characters.csv"}},"sourceStats":{"recordCount":1.0}}}
{"type":"LOG","log":{"level":"INFO","message":"Read 1 records from simpsons_characters stream"}}
{"type":"LOG","log":{"level":"INFO","message":"Marking stream simpsons_characters as STOPPED"}}
{"type":"TRACE","trace":{"type":"STREAM_STATUS","emitted_at":1730318469031.434,"stream_status":{"stream_descriptor":{"name":"simpsons_characters"},"status":"COMPLETE"}}}
{"type":"LOG","log":{"level":"INFO","message":"Finished syncing simpsons_characters"}}
{"type":"LOG","log":{"level":"INFO","message":"SourceSFTPBulk runtimes:\nSyncing stream simpsons_characters 0:01:19.319735\nSyncing stream simpsons_locations 0:01:51.654029"}}
{"type":"LOG","log":{"level":"INFO","message":"Finished syncing SourceSFTPBulk"}}
{"type":"TRACE","trace":{"type":"ANALYTICS","emitted_at":1730318469034.845,"analytics":{"type":"file-cdk-csv-stream-count","value":"2"}}}
```